roofit/roofitcore/src/RooAbsData.cxx

```diff
@@ -2330,8 +2331,43 @@ Bool_t RooAbsData::hasFilledCache() const
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+/// give access to internal data TTree (if it exists)
+/// RooDataSet keeps ownership, RooFit is responsible for deleting.
 
 const TTree* RooAbsData::tree() const
 {
-  return _dstore->tree() ;
+  if (dynamic_cast<const RooTreeDataStore*>(_dstore)) {
+    return _dstore->tree();
+  } else {
+    coutW(InputArguments) << "RooAbsData::tree(" << GetName() << ") WARNING: is not of StorageType::Tree. "
+             << "Use export_tree() instead or convert to tree storage." << endl ;
+    return (TTree*)nullptr;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// clone internal data TTree or create it from vector-based storage
+/// transfers ownership, user is responsible for deleting.
+
+TTree* RooAbsData::export_tree() const
+{
+  if (dynamic_cast<const RooTreeDataStore*>(_dstore)) {
+    return const_cast<TTree*>(_dstore->tree())->CloneTree();
+  } else {
+    RooTreeDataStore buffer(GetName(), GetTitle(), *get(), *_dstore);
+    TTree* t = &(buffer.tree());
+    return (TTree*)(t->CloneTree());
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Convert vector-based storage to tree-based storage
+
+void RooAbsData::convertToTreeStore()
+{
+  if (nullptr==dynamic_cast<RooTreeDataStore*>(_dstore)) {
+    RooTreeDataStore* newStore = new RooTreeDataStore(GetName(), GetTitle(), *get(), *_dstore);
+    delete _dstore;
```

Comment:
The unique_ptr was introduced in C++11 and this code is much older. That's why we should update to newer language features when we are working in the area, even though it might look unnecessary.